### PR TITLE
[ORCA] Enhance Continuous Training with TF2 PySpark Backend

### PR DESCRIPTION
**Pull request by sgwhat**
## Description
Fix for https://github.com/intel-analytics/BigDL/issues/7268, which needs to enhance the continuous training pipeline when `model_dir` is None.

This PR only supports TensorFlow 2.10.0 and previous versions, for 2.11.0 version, we will complete the support in PR https://github.com/intel-analytics/BigDL/pull/7423 later.

### 1. Why the change?
For the previous implement, the optimizer weight is not supported to save, which could cause a gap when training with loading a saved model.

### 2. User API changes
N/A

### 3. Summary of the change 
Support return `optimizer_weights` after the training step, and re-create the model by loading `optimizer_weights` when calling `get_model` (same as `model_weights` does).

### 4. How to test?
- [x] Unit test

### CODE REVIEW

**Comment by sgwhat** [comment]
Pending adding a new UT.

File: python/orca/src/bigdl/orca/learn/tf2/pyspark_estimator.py [hunk 1]
```diff
@@ -635,8 +643,15 @@ def get_model(self,
...
-            model.set_weights(self.model_weights)
+        if set_weights:
+            if self.model_weights is not None:
+                model.set_weights(self.model_weights)
+            if self.optimizer_weights is not None:
+                import tensorflow as tf
+                grad_vars = model.trainable_weights
+                zero_grads = [tf.zeros_like(w) for w in grad_vars]
+                model.optimizer.apply_gradients(zip(zero_grads, grad_vars))
+                model.optimizer.set_weights(self.optimizer_weights)
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/tf2/pyspark_estimator.py:654]: shall we set optimizer weights before setting model weights?

Referred to this post: https://stackoverflow.com/questions/49503748/save-and-load-model-optimizer-state
image

File: python/orca/src/bigdl/orca/learn/tf2/spark_runner.py [hunk 2]
```diff
@@ -260,6 +262,15 @@ def __init__(self,
...
                                     self.model.set_weights(self.model_weights.value)
+                                if self.optimizer_weights:
+                                    def load_opt_weights():
+                                        grad_vars = self.model.trainable_weights
+                                        zero_grads = [tf.zeros_like(w) for w in grad_vars]
+                                        self.model.optimizer.apply_gradients(
+                                            zip(zero_grads, grad_vars))
+                                        self.model.optimizer.set_weights(
+                                            self.optimizer_weights.value)
+                                    self.strategy.run(load_opt_weights)
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/tf2/spark_runner.py:273]: what's the difference between calling strategy.run and directly calling this like setting model weights?

Comment (sgwhat) [reply python/orca/src/bigdl/orca/learn/tf2/spark_runner.py:273]: it's not allowed to call `model.optimizer.set_weights` directly inside the strategy scope.

**Review by hkvision — APPROVED** [review]

**Comment by sgwhat** [comment]
test by NCF example:
1. tf_train_spark_df.py lr [0.1, 0.05]
2. tf_resume_train_spark_df.py lr [0.05, 0.025]

**Comment by hkvision** [comment]
> test by NCF example:
> 
> 1. tf_train_spark_df.py lr [0.1, 0.05]
> 2. tf_resume_train_spark_df.py lr [0.05, 0.025]

Seems the first epoch lr of resume train is wrong, need to check and verify this.

**Comment by zpeng1898** [comment]
Test `tf_resume_train_spark_dataframe.py --lr_scheduler --backend=ray` showing `AttributeError: 'Adam' object has no attribute 'get_weights'`

error message:
```
2023-02-22 18:18:21,369 Thread-5 ERROR Unable to create Appender of type File
cls.getname: com.intel.analytics.bigdl.dllib.utils.python.api.Sample
BigDLBasePickler registering: bigdl.dllib.utils.common  Sample
cls.getname: com.intel.analytics.bigdl.dllib.utils.python.api.EvaluatedResult
BigDLBasePickler registering: bigdl.dllib.utils.common  EvaluatedResult
cls.getname: com.intel.analytics.bigdl.dllib.utils.python.api.JTensor
BigDLBasePickler registering: bigdl.dllib.utils.common  JTensor
cls.getname: com.intel.analytics.bigdl.dllib.utils.python.api.JActivity
BigDLBasePickler registering: bigdl.dllib.utils.common  JActivity
2023-02-22 18:18:31.424423: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2023-02-22 18:18:31.424508: W tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:265] failed call to cuInit: UNKNOWN ERROR (303)
2023-02-22 18:18:31.424554: I tensorflow/compiler/xla/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (Almaren-Node-107): /proc/driver/nvidia/version does not exist
2023-02-22 18:18:31.425111: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Traceback (most recent call last):
  File "tf_resume_train_spark_dataframe.py", line 44, in <module>
    est.load(os.path.join(args.model_dir, "NCF_model"))
  File "/home/kai/anaconda3/envs/py378/lib/python3.7/site-packages/bigdl/orca/learn/tf2/pyspark_estimator.py", line 623, in load
    self.optimizer_weights = model.optimizer.get_weights()
AttributeError: 'Adam' object has no attribute 'get_weights'
Stopping orca context
```

**Comment by hkvision** [comment]
tf 2.11 hasn't been supported yet. Use lower versions e.g. 2.9

**Comment by zpeng1898** [comment]
test NCF example with `tensorflow==2.9.0`:
- `python tf_train_spark_dataframe.py  --backend spark --lr_scheduler`: `lr: [0.01, 0.005]`
- `python tf_resume_train_spark_dataframe.py --backend spark --lr_scheduler`: `lr: [0.005, 0.0025]`

- `python tf_train_spark_dataframe.py --backend ray --lr_scheduler`: `lr: [0.01, 0.005]`
- `python tf_resume_train_spark_dataframe.py --backend ray --lr_scheduler`: 
```
2023-02-22 19:27:08.300086: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Traceback (most recent call last):
  File "tf_resume_train_spark_dataframe.py", line 44, in <module>
    est.load(os.path.join(args.model_dir, "NCF_model"))
  File "/home/kai/anaconda3/envs/py378/lib/python3.7/site-packages/bigdl/orca/learn/tf2/pyspark_estimator.py", line 621, in load
    model = load_model(**self.load_params)
  File "/home/kai/anaconda3/envs/py378/lib/python3.7/site-packages/bigdl/orca/learn/utils.py", line 784, in load_model
    compile=compile
  File "/home/kai/anaconda3/envs/py378/lib/python3.7/site-packages/keras/utils/traceback_utils.py", line 67, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/home/kai/anaconda3/envs/py378/lib/python3.7/site-packages/keras/optimizers/optimizer_experimental/optimizer.py", line 94, in _process_kwargs
    raise TypeError(f"{k} is not a valid argument, kwargs should be empty "
TypeError: weight_decay is not a valid argument, kwargs should be empty  for `optimizer_experimental.Optimizer`.
Stopping orca context
```

**Pull request merged**